aws/resource_aws_transfer_user_test.go

```diff
@@ -125,6 +125,32 @@ func TestAccAWSTransferUser_disappears(t *testing.T) {
 	})
 }
 
+func TestAccAWSTransferUserName_validation(t *testing.T) {
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSTransferUserDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config:      testAccAWSTransferUserName_validation("!@#$%^"),
+				ExpectError: regexp.MustCompile(`"user_name" can only contain alphanumeric characters, underscores, and hyphens`),
+			},
+			{
+				Config:      testAccAWSTransferUserName_validation(acctest.RandString(2)),
+				ExpectError: regexp.MustCompile(`"user_name" must be at least 3 characters`),
+			},
+			{
+				Config:      testAccAWSTransferUserName_validation(acctest.RandString(33)),
+				ExpectError: regexp.MustCompile(`"user_name" cannot be more than 32 characters`),
+			},
+			{
+				Config:      testAccAWSTransferUserName_validation("-abcdef"),
+				ExpectError: regexp.MustCompile(`"user_name" cannot begin with a hyphen`),
+			},
```

Comment:
That's cool! You never test that your validation function itself "passes" though.
You might want to add that step.

Reply:
Not sure I understand, is that not what I had already done via below?
```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSTransferUserName_validation'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestAccAWSTransferUserName_validation -timeout 120m
=== RUN   TestAccAWSTransferUserName_validation
=== PAUSE TestAccAWSTransferUserName_validation
=== CONT  TestAccAWSTransferUserName_validation
--- PASS: TestAccAWSTransferUserName_validation (1.64s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	1.681s

...
```
I'll be sure to add whatever else is required.

Reply:
What I meant is that you've tested all the errors but you have no test saying that the name is correct.
They're all expecting errors.

Reply:
Understood, I'll take care of that now.

Reply:
@charlyx could you take another look at this?